File: tests/fortran/assumed_types/end_to_end/fixtures/native/assumed_type_mutation.f90
subroutine bump_address(x) bind(C)
  type(*), dimension(*), intent(inout) :: x
end subroutine

subroutine bump_descriptor(x) bind(C)
  type(*), dimension(:), intent(inout) :: x
end subroutine

subroutine bump_scalar(x) bind(C)
  type(*), intent(inout) :: x
end subroutine

File: tests/fortran/assumed_types/end_to_end/fixtures/native/assumed_type_mutation.c
#include <stdint.h>
#include <ISO_Fortran_binding.h>

void bump_address(void *x)
{
    ((int64_t *)x)[0] += 1;
}

void bump_descriptor(CFI_cdesc_t *x)
{
    int64_t *first = (int64_t *)x->base_addr;
    int64_t *last = (int64_t *)((char *)x->base_addr + (x->dim[0].extent - 1) * x->dim[0].sm);
    *first += 1;
    *last += 1;
}

void bump_scalar(void *x)
{
    *(int64_t *)x += 1;
}
